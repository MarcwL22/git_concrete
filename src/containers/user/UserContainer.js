import React, { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
// Actions
import { userOperations } from './duck';
import { homeOperations } from '../home/duck';
// Components
import UserLoading from './components/UserLoading';
import UserBackButton from './components/UserBackButton';
import UserModalRepository from './components/UserModalRepository';
import UserError from './components/UserError';
import UserInfo from './components/UserInfo';
import UserRepositories from './components/UserRepositories';
// Variables
import { BREAKPOINTS } from '../../assets/variables';
// Util
import { orderArrByName } from '../../util';
// Mocks
// import userMock from './userMock.json';
// import reposMock from './repositoriesMock.json';
// import repoMock from './repository.json';

const MainContainer = styled.div`
  display: flex;
  @media (max-width: ${BREAKPOINTS.phone}) {
    flex-direction: column;
  }
`;

export class UserContainer extends Component {
  state = {
    orderByName: false,
    orderByStar: false
  };

  componentDidMount() {
    const { user } = this.props.match.params;
    this.props.fetchUser(user);
    this.props.fetchUserRepositories(user);
  }

  backHome = () => this.props.history.push('/');

  onRepoClick = fullName => this.props.fetchRepositoryDetail(fullName);

  closeModal = () => this.props.handleModal();

  onPressOrderByStar = () => {
    this.setState({ orderByStar: !this.state.orderByStar, orderByName: false });
  };

  onPressOrderByName = () => {
    this.setState({ orderByName: !this.state.orderByName, orderByStar: false });
  };

  orderArr = array => {
    const { orderByStar } = this.state;
    if (orderByStar) return array.sort((a, b) => b.stargazers_count - a.stargazers_count);
    return orderArrByName(array);
  };

  renderContent = () => {
    const { homeReducer, loadingRepos, errorRepos, repositories } = this.props;
    const { loading, error, user } = homeReducer;
    const { orderByStar } = this.state;
    if (loading && loadingRepos) {
      return <UserLoading />;
    } else if (errorRepos && error) {
      return <UserError />;
    } else {
      return (
        <React.Fragment>
          <UserInfo userData={user} />
          <UserRepositories
            repositories={this.orderArr(repositories)}
            starOrdered={orderByStar}
            onPressOrderStar={this.onPressOrderByStar}
            onRepositoryPress={this.onRepoClick}
          />
        </React.Fragment>
      );
    }
  };

  render() {
    const { modalRepoOpen, errorRepoDetail, loadingRepoDetail, repository } = this.props;
    return (
      <React.Fragment>
        <UserBackButton onPress={this.backHome} />
        <MainContainer>{this.renderContent()}</MainContainer>
        <UserModalRepository
          repo={repository}
          modalOpen={modalRepoOpen}
          onClose={this.closeModal}
          error={errorRepoDetail}
          loading={loadingRepoDetail}
        />
      </React.Fragment>
    );
  }
}

const mapStateToProps = ({ homeReducer, userReducer }) => {
  return { ...userReducer, homeReducer };
};

export default connect(
  mapStateToProps,
  { ...userOperations, ...homeOperations }
)(UserContainer);
