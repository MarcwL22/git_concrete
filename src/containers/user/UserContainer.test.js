import React from 'react';
import { shallow, mount } from 'enzyme';
import { UserContainer } from './UserContainer';

describe('UserContainer', () => {
  let props = {
    homeReducer: {
      loading: false,
      error: false,
      user: {}
    },
    loadingRepos: false,
    loadingRepoDetail: false,
    errorRepos: false,
    errorRepoDetail: false,
    modalRepoOpen: false,
    repositories: [],
    repository: {}
  };
  let userContainer = shallow(<UserContainer {...props} />);
  it('should renders properly', () => {
    expect(userContainer).toMatchSnapshot();
  });
  describe.skip('should when mounted', () => {
    const mockFetchUser = jest.fn();
    const mockFetchUserRepositories = jest.fn();
    beforeEach(() => {
      props.fetchUser = mockFetchUser;
      props.fetchUserRepositories = mockFetchUserRepositories;
      console.log('props: ', props);
      userContainer = mount(<UserContainer {...props} />);
    });
    it('should dispatches the `fetchUser()` method it receives from props', () => {
      expect(mockFetchUser).toHaveBeenLastCalledWith('Marcwl22');
    });
  });
});
