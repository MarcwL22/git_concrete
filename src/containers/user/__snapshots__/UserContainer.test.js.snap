// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserContainer should renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserContainer
    errorRepoDetail={false}
    errorRepos={false}
    homeReducer={
        Object {
            "error": false,
            "loading": false,
            "user": Object {},
          }
    }
    loadingRepoDetail={false}
    loadingRepos={false}
    modalRepoOpen={false}
    repositories={Array []}
    repository={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <UserBackButton
          onPress={[Function]}
/>,
        <styled.div>
          <Unknown>
                    <UserInfo
                              userData={Object {}}
                    />
                    <UserRepositories
                              onPressOrderStar={[Function]}
                              onRepositoryPress={[Function]}
                              repositories={Array []}
                              starOrdered={false}
                    />
          </Unknown>
</styled.div>,
        <UserModalRepository
          error={false}
          loading={false}
          modalOpen={false}
          onClose={[Function]}
          repo={Object {}}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onPress": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Unknown>
            <UserInfo
                        userData={Object {}}
            />
            <UserRepositories
                        onPressOrderStar={[Function]}
                        onRepositoryPress={[Function]}
                        repositories={Array []}
                        starOrdered={false}
            />
</Unknown>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <UserInfo
                userData={Object {}}
/>,
              <UserRepositories
                onPressOrderStar={[Function]}
                onRepositoryPress={[Function]}
                repositories={Array []}
                starOrdered={false}
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "userData": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "onPressOrderStar": [Function],
                "onRepositoryPress": [Function],
                "repositories": Array [],
                "starOrdered": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": Symbol(react.fragment),
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "error": false,
          "loading": false,
          "modalOpen": false,
          "onClose": [Function],
          "repo": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <UserBackButton
            onPress={[Function]}
/>,
          <styled.div>
            <Unknown>
                        <UserInfo
                                    userData={Object {}}
                        />
                        <UserRepositories
                                    onPressOrderStar={[Function]}
                                    onRepositoryPress={[Function]}
                                    repositories={Array []}
                                    starOrdered={false}
                        />
            </Unknown>
</styled.div>,
          <UserModalRepository
            error={false}
            loading={false}
            modalOpen={false}
            onClose={[Function]}
            repo={Object {}}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onPress": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Unknown>
              <UserInfo
                            userData={Object {}}
              />
              <UserRepositories
                            onPressOrderStar={[Function]}
                            onRepositoryPress={[Function]}
                            repositories={Array []}
                            starOrdered={false}
              />
</Unknown>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <UserInfo
                  userData={Object {}}
/>,
                <UserRepositories
                  onPressOrderStar={[Function]}
                  onRepositoryPress={[Function]}
                  repositories={Array []}
                  starOrdered={false}
/>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "userData": Object {},
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "onPressOrderStar": [Function],
                  "onRepositoryPress": [Function],
                  "repositories": Array [],
                  "starOrdered": false,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": Symbol(react.fragment),
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "error": false,
            "loading": false,
            "modalOpen": false,
            "onClose": [Function],
            "repo": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "disableLifecycleMethods": true,
    "hydrateIn": undefined,
  },
}
`;
